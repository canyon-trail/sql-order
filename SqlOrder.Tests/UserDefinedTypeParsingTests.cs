using Snapshooter.Xunit;
using SqlOrder.AstTypes;

namespace SqlOrder.Tests;

public sealed class UserDefinedTypeParsingTests
{
    [Fact]
    public void SimpleScalarType()
    {
        var sql = @"
            create type derp from varchar(5) not null;
        ";

        var parser = new SqlParser();

        var results = parser.Parse(sql);

        results.Should().BeEquivalentTo(new[] {
            new UserDefinedTypeDefinition(
                new ObjectName("dbo", "derp"),
                Dependency.EmptyArray
            ),
        });
    }

    [Fact]
    public void SimpleTableType()
    {
        var sql = @"
            create type derp as table (
                thingId int not null,
                name text not null
            );
        ";

        var parser = new SqlParser();

        var results = parser.Parse(sql);

        results.Should().BeEquivalentTo(new[] {
            new UserDefinedTypeDefinition(
                new ObjectName("dbo", "derp"),
                Dependency.EmptyArray
            ),
        });
    }

    [Fact]
    public void TableWithCustomTypeColumn()
    {
        var sql = @"
            create type derp as table (
                thingId int not null,
                name text not null,
                otherthing varchar(3) null,
                dep otherschema.deptype null
            );
        ";

        SqlParser.ParseInternal(sql).Simplify().Should().MatchSnapshot();

        var parser = new SqlParser();

        var results = parser.Parse(sql);

        results.Should().BeEquivalentTo(new[] {
            new UserDefinedTypeDefinition(
                new ObjectName("dbo", "derp"),
                Dependency.ArrayOf(
                    new Dependency(new ObjectName("otherschema", "deptype"), DependencyKind.UserDefinedType)
                )
            ),
        });
    }
}
