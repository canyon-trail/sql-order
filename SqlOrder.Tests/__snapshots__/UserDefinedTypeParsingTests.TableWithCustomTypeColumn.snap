{
  "Name": "SqlScript",
  "Children": [
    {
      "Name": "SqlBatch",
      "Children": [
        {
          "Name": "SqlCreateUserDefinedTableTypeStatement",
          "Children": [
            {
              "Name": "OnePartObjectIdentifier",
              "Children": [
                {
                  "Name": "SqlIdentifier",
                  "Children": []
                }
              ]
            },
            {
              "Name": "SqlTableDefinition",
              "Children": [
                {
                  "Name": "SqlColumnDefinition",
                  "Children": [
                    {
                      "Name": "SqlIdentifier",
                      "Children": []
                    },
                    {
                      "Name": "SqlDataTypeSpecification",
                      "Children": [
                        {
                          "Name": "SqlDataType",
                          "Children": [
                            {
                              "Name": "OnePartObjectIdentifier",
                              "Children": [
                                {
                                  "Name": "SqlIdentifier",
                                  "Children": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Name": "TypeOnlyConstraint",
                      "Children": []
                    }
                  ]
                },
                {
                  "Name": "SqlColumnDefinition",
                  "Children": [
                    {
                      "Name": "SqlIdentifier",
                      "Children": []
                    },
                    {
                      "Name": "SqlDataTypeSpecification",
                      "Children": [
                        {
                          "Name": "SqlDataType",
                          "Children": [
                            {
                              "Name": "OnePartObjectIdentifier",
                              "Children": [
                                {
                                  "Name": "SqlIdentifier",
                                  "Children": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Name": "TypeOnlyConstraint",
                      "Children": []
                    }
                  ]
                },
                {
                  "Name": "SqlColumnDefinition",
                  "Children": [
                    {
                      "Name": "SqlIdentifier",
                      "Children": []
                    },
                    {
                      "Name": "SqlDataTypeSpecification",
                      "Children": [
                        {
                          "Name": "SqlDataType",
                          "Children": [
                            {
                              "Name": "OnePartObjectIdentifier",
                              "Children": [
                                {
                                  "Name": "SqlIdentifier",
                                  "Children": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Name": "TypeOnlyConstraint",
                      "Children": []
                    }
                  ]
                },
                {
                  "Name": "SqlColumnDefinition",
                  "Children": [
                    {
                      "Name": "SqlIdentifier",
                      "Children": []
                    },
                    {
                      "Name": "SqlDataTypeSpecification",
                      "Children": [
                        {
                          "Name": "SqlDataType",
                          "Children": [
                            {
                              "Name": "TwoPartObjectIdentifier",
                              "Children": [
                                {
                                  "Name": "SqlIdentifier",
                                  "Children": []
                                },
                                {
                                  "Name": "SqlIdentifier",
                                  "Children": []
                                }
                              ]
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Name": "TypeOnlyConstraint",
                      "Children": []
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
